rust/otap-dataflow/crates/admin/src/telemetry.rs

```diff
@@ -0,0 +1,866 @@
+
+    out
+}
+
+fn format_prometheus_text(
+    registry: &MetricsRegistryHandle,
+    reset: bool,
+    timestamp_millis: Option<i64>,
+) -> String {
+    let mut out = String::new();
+    let ts_suffix = timestamp_millis
+        .map(|ms| format!(" {}", ms))
+        .unwrap_or_default();
+    let mut seen: HashSet<String> = HashSet::new();
+
+    let mut visit = |descriptor: &'static MetricsDescriptor,
+                     attributes: &dyn AttributeSetHandler,
+                     metrics_iter: MetricsIterator<'_>| {
+        // Render labels from attributes + set label
+        let mut base_labels = String::new();
+        if !descriptor.name.is_empty() {
+            let _ = write!(
+                &mut base_labels,
+                "set=\"{}\"",
+                escape_prom_label_value(descriptor.name)
+            );
+        }
+        for (key, value) in attributes.iter_attributes() {
+            if !base_labels.is_empty() {
+                base_labels.push(',');
+            }
+            let _ = write!(
+                &mut base_labels,
+                "{}=\"{}\"",
+                sanitize_prom_label_key(key),
+                escape_prom_label_value(&value.to_string_value())
+            );
+        }
+
+        for (field, value) in metrics_iter {
+            let metric_name = sanitize_prom_metric_name(field.name);
+
+            // HELP/TYPE once per metric name
+            if seen.insert(metric_name.clone()) {
+                if !field.brief.is_empty() {
+                    let _ = writeln!(
+                        &mut out,
+                        "# HELP {} {}",
+                        metric_name,
+                        escape_prom_help(field.brief)
+                    );
+                }
+                let prom_type = match field.instrument {
+                    Instrument::Counter => "counter",
+                    Instrument::UpDownCounter => "gauge",
+                    Instrument::Gauge => "gauge",
+                    Instrument::Histogram => "gauge",
+                };
+                let _ = writeln!(&mut out, "# TYPE {} {}", metric_name, prom_type);
+            }
+
+            if base_labels.is_empty() {
+                let _ = writeln!(&mut out, "{} {}{}", metric_name, value, ts_suffix);
+            } else {
+                let _ = writeln!(
+                    &mut out,
+                    "{}{{{}}} {}{}",
+                    metric_name, base_labels, value, ts_suffix
+                );
+            }
+        }
+    };
+
+    if reset {
+        registry.visit_metrics_and_reset(|d, a, m| visit(d, a, m));
+    } else {
+        registry.visit_current_metrics(|d, a, m| visit(d, a, m));
+    }
+
+    out
+}
+
+fn escape_lp_measurement(s: &str) -> String {
+    s.replace(',', "\\,").replace(' ', "\\ ")
+}
+
+fn escape_lp_tag_key(s: &str) -> String {
+    s.replace(',', "\\,")
+        .replace(' ', "\\ ")
+        .replace('=', "\\=")
+}
+
+fn escape_lp_tag_value(s: &str) -> String {
```

Comment:
`escape_lp_tag_value` and  `escape_lp_tag_key` methods seem to be identical in implementation. Could we just use one method instead? Maybe call it `escape_lp_tag`?

Reply:
Removed code duplication